src/fs/data_cluster_contents_info.hh

```diff
@@ -0,0 +1,83 @@
+/*
+ * This file is open source software, licensed to you under the terms
+ * of the Apache License, Version 2.0 (the "License").  See the NOTICE file
+ * distributed with this work for additional information regarding copyright
+ * ownership.  You may not use this file except in compliance with the License.
+ *
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+/*
+ * Copyright (C) 2020 ScyllaDB
+ */
+
+#pragma once
+
+#include "fs/cluster.hh"
+#include "fs/inode.hh"
+#include "fs/inode_info.hh"
+#include "fs/units.hh"
+
+#include <map>
+
+namespace seastar::fs {
+
+class data_cluster_contents_info {
+public:
+    // Each cluster_data_vec corresponds to an inode_data_vec
+    struct cluster_data_vec {
+        inode_t data_owner;
+        file_range data_range;
+    };
+
+private:
+    size_t _valid_data_size = 0;
+    bool _read_only = false;
+    std::map<disk_offset_t, cluster_data_vec> _data;
+
+public:
+    const std::map<disk_offset_t, cluster_data_vec>& get_data() const noexcept {
+        return _data;
+    }
+
+    bool is_empty() const noexcept {
+        return _valid_data_size == 0;
+    }
+
+    void finished_writing() noexcept {
+        _read_only = true;
+    }
+
+    bool can_compact() const noexcept {
+        return _read_only;
+    }
+
+    void add_data(disk_offset_t disk_offset, inode_t inode, file_range data_range) noexcept {
+        _data.emplace(disk_offset, cluster_data_vec {inode, data_range});
+        _valid_data_size += data_range.size();
+    }
+
+    // Removes data_vec stored at disk_offset and adds remaining data_
+    void cut_data(disk_offset_t disk_offset, file_range former_range, file_range new_left_range, file_range new_right_range) {
```

Comment:
Looks good, but it took me a while to go over some example corner cases and convince myself that there are no obvious bugs. Do we have a unit test for the case when *both* left and right remains, i.e. when we punch a hole in a file? I saw truncate tests, but unless I'm mistaken, they only cover cutting data from the end.